File: pdu/python/hako_mavlink_msgs/pdu_conv_HakoSERVO_OUTPUT_RAW.py

import struct
from .pdu_pytype_HakoSERVO_OUTPUT_RAW import HakoSERVO_OUTPUT_RAW
from ..pdu_utils import *
from .. import binary_io

# dependencies for the generated Python class



def pdu_to_py_HakoSERVO_OUTPUT_RAW(binary_data: bytearray) -> HakoSERVO_OUTPUT_RAW:
    py_obj = HakoSERVO_OUTPUT_RAW()
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        raise ValueError("Invalid PDU binary data: MetaData not found or corrupted")
    binary_read_recursive_HakoSERVO_OUTPUT_RAW(meta, binary_data, py_obj, binary_io.PduMetaData.PDU_META_DATA_SIZE)
    return py_obj


def binary_read_recursive_HakoSERVO_OUTPUT_RAW(meta: binary_io.PduMetaData, binary_data: bytearray, py_obj: HakoSERVO_OUTPUT_RAW, base_off: int):
    # array_type: single 
    # data_type: primitive 
    # member_name: time_usec 
    # type_name: uint32 
    # offset: 0 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 0, 4)
    py_obj.time_usec = binary_io.binTovalue("uint32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: port 
    # type_name: uint8 
    # offset: 4 size: 1 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 4, 1)
    py_obj.port = binary_io.binTovalue("uint8", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo1_raw 
    # type_name: uint16 
    # offset: 6 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 6, 2)
    py_obj.servo1_raw = binary_io.binTovalue("uint16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo2_raw 
    # type_name: uint16 
    # offset: 8 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 8, 2)
    py_obj.servo2_raw = binary_io.binTovalue("uint16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo3_raw 
    # type_name: uint16 
    # offset: 10 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 10, 2)
    py_obj.servo3_raw = binary_io.binTovalue("uint16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo4_raw 
    # type_name: uint16 
    # offset: 12 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 12, 2)
    py_obj.servo4_raw = binary_io.binTovalue("uint16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo5_raw 
    # type_name: uint16 
    # offset: 14 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 14, 2)
    py_obj.servo5_raw = binary_io.binTovalue("uint16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo6_raw 
    # type_name: uint16 
    # offset: 16 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 16, 2)
    py_obj.servo6_raw = binary_io.binTovalue("uint16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo7_raw 
    # type_name: uint16 
    # offset: 18 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 18, 2)
    py_obj.servo7_raw = binary_io.binTovalue("uint16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo8_raw 
    # type_name: uint16 
    # offset: 20 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 20, 2)
    py_obj.servo8_raw = binary_io.binTovalue("uint16", bin)
    
    return py_obj


def py_to_pdu_HakoSERVO_OUTPUT_RAW(py_obj: HakoSERVO_OUTPUT_RAW) -> bytearray:
    binary_data = bytearray()
    base_allocator = DynamicAllocator(False)
    bw_container = BinaryWriterContainer(binary_io.PduMetaData())
    binary_write_recursive_HakoSERVO_OUTPUT_RAW(0, bw_container, base_allocator, py_obj)

    # メタデータの設定
    total_size = base_allocator.size() + bw_container.heap_allocator.size() + binary_io.PduMetaData.PDU_META_DATA_SIZE
    bw_container.meta.total_size = total_size
    bw_container.meta.heap_off = binary_io.PduMetaData.PDU_META_DATA_SIZE + base_allocator.size()

    # binary_data のサイズを total_size に調整
    if len(binary_data) < total_size:
        binary_data.extend(bytearray(total_size - len(binary_data)))
    elif len(binary_data) > total_size:
        del binary_data[total_size:]

    # メタデータをバッファにコピー
    binary_io.writeBinary(binary_data, 0, bw_container.meta.to_bytes())

    # 基本データをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.base_off, base_allocator.to_array())

    # ヒープデータをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.heap_off, bw_container.heap_allocator.to_array())

    return binary_data

def binary_write_recursive_HakoSERVO_OUTPUT_RAW(parent_off: int, bw_container: BinaryWriterContainer, allocator, py_obj: HakoSERVO_OUTPUT_RAW):
    # array_type: single 
    # data_type: primitive 
    # member_name: time_usec 
    # type_name: uint32 
    # offset: 0 size: 4 
    # array_len: 1
    type = "uint32"
    off = 0

    
    bin = binary_io.typeTobin(type, py_obj.time_usec)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: port 
    # type_name: uint8 
    # offset: 4 size: 1 
    # array_len: 1
    type = "uint8"
    off = 4

    
    bin = binary_io.typeTobin(type, py_obj.port)
    bin = get_binary(type, bin, 1)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo1_raw 
    # type_name: uint16 
    # offset: 6 size: 2 
    # array_len: 1
    type = "uint16"
    off = 6

    
    bin = binary_io.typeTobin(type, py_obj.servo1_raw)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo2_raw 
    # type_name: uint16 
    # offset: 8 size: 2 
    # array_len: 1
    type = "uint16"
    off = 8

    
    bin = binary_io.typeTobin(type, py_obj.servo2_raw)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo3_raw 
    # type_name: uint16 
    # offset: 10 size: 2 
    # array_len: 1
    type = "uint16"
    off = 10

    
    bin = binary_io.typeTobin(type, py_obj.servo3_raw)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo4_raw 
    # type_name: uint16 
    # offset: 12 size: 2 
    # array_len: 1
    type = "uint16"
    off = 12

    
    bin = binary_io.typeTobin(type, py_obj.servo4_raw)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo5_raw 
    # type_name: uint16 
    # offset: 14 size: 2 
    # array_len: 1
    type = "uint16"
    off = 14

    
    bin = binary_io.typeTobin(type, py_obj.servo5_raw)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo6_raw 
    # type_name: uint16 
    # offset: 16 size: 2 
    # array_len: 1
    type = "uint16"
    off = 16

    
    bin = binary_io.typeTobin(type, py_obj.servo6_raw)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo7_raw 
    # type_name: uint16 
    # offset: 18 size: 2 
    # array_len: 1
    type = "uint16"
    off = 18

    
    bin = binary_io.typeTobin(type, py_obj.servo7_raw)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: servo8_raw 
    # type_name: uint16 
    # offset: 20 size: 2 
    # array_len: 1
    type = "uint16"
    off = 20

    
    bin = binary_io.typeTobin(type, py_obj.servo8_raw)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    

if __name__ == "__main__":
    import sys
    import json

    def print_usage():
        print(f"Usage: python -m pdu.python.pdu_conv_HakoSERVO_OUTPUT_RAW <read|write> [args...]")
        print(f"  read <input_binary_file> <output_json_file>")
        print(f"  write <input_json_file> <output_binary_file>")

    if len(sys.argv) < 2:
        print_usage()
        sys.exit(1)

    command = sys.argv[1]

    if command == "read":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)
        
        binary_filepath = sys.argv[2]
        output_json_filepath = sys.argv[3]

        with open(binary_filepath, "rb") as f:
            binary_data = bytearray(f.read())
        
        py_obj = pdu_to_py_HakoSERVO_OUTPUT_RAW(binary_data)
        
        with open(output_json_filepath, "w") as f:
            f.write(py_obj.to_json())

    elif command == "write":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)

        input_json_filepath = sys.argv[2]
        output_binary_filepath = sys.argv[3]

        with open(input_json_filepath, "r") as f:
            json_str = f.read()
        
        py_obj = HakoSERVO_OUTPUT_RAW.from_json(json_str)
        
        binary_data = py_to_pdu_HakoSERVO_OUTPUT_RAW(py_obj)

        with open(output_binary_filepath, "wb") as f:
            f.write(binary_data)

    else:
        print(f"Unknown command: {command}")
        print_usage()
        sys.exit(1)

File: pdu/javascript/hako_mavlink_msgs/pdu_conv_HakoSERVO_OUTPUT_RAW.js
import * as PduUtils from '../pdu_utils.js';
import { HakoSERVO_OUTPUT_RAW } from './pdu_jstype_HakoSERVO_OUTPUT_RAW.js';


/**
 * Deserializes a binary PDU into a HakoSERVO_OUTPUT_RAW object.
 * @param {ArrayBuffer} binary_data
 * @returns { HakoSERVO_OUTPUT_RAW }
 */
export function pduToJs_HakoSERVO_OUTPUT_RAW(binary_data) {
    const js_obj = new HakoSERVO_OUTPUT_RAW();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_HakoSERVO_OUTPUT_RAW(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_HakoSERVO_OUTPUT_RAW(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: time_usec, type: uint32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 0, 4);
        js_obj.time_usec = PduUtils.binToValue("uint32", bin);
    }
    
    // member: port, type: uint8 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 4, 1);
        js_obj.port = PduUtils.binToValue("uint8", bin);
    }
    
    // member: servo1_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 6, 2);
        js_obj.servo1_raw = PduUtils.binToValue("uint16", bin);
    }
    
    // member: servo2_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 8, 2);
        js_obj.servo2_raw = PduUtils.binToValue("uint16", bin);
    }
    
    // member: servo3_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 10, 2);
        js_obj.servo3_raw = PduUtils.binToValue("uint16", bin);
    }
    
    // member: servo4_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 12, 2);
        js_obj.servo4_raw = PduUtils.binToValue("uint16", bin);
    }
    
    // member: servo5_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 14, 2);
        js_obj.servo5_raw = PduUtils.binToValue("uint16", bin);
    }
    
    // member: servo6_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 16, 2);
        js_obj.servo6_raw = PduUtils.binToValue("uint16", bin);
    }
    
    // member: servo7_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 18, 2);
        js_obj.servo7_raw = PduUtils.binToValue("uint16", bin);
    }
    
    // member: servo8_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 20, 2);
        js_obj.servo8_raw = PduUtils.binToValue("uint16", bin);
    }
    
    return js_obj;
}

/**
 * Serializes a HakoSERVO_OUTPUT_RAW object into a binary PDU.
 * @param { HakoSERVO_OUTPUT_RAW } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_HakoSERVO_OUTPUT_RAW(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_HakoSERVO_OUTPUT_RAW(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_HakoSERVO_OUTPUT_RAW(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: time_usec, type: uint32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint32", js_obj.time_usec, 4);
        allocator.add(bin, parent_off + 0);
    }
    
    // member: port, type: uint8 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint8", js_obj.port, 1);
        allocator.add(bin, parent_off + 4);
    }
    
    // member: servo1_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint16", js_obj.servo1_raw, 2);
        allocator.add(bin, parent_off + 6);
    }
    
    // member: servo2_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint16", js_obj.servo2_raw, 2);
        allocator.add(bin, parent_off + 8);
    }
    
    // member: servo3_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint16", js_obj.servo3_raw, 2);
        allocator.add(bin, parent_off + 10);
    }
    
    // member: servo4_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint16", js_obj.servo4_raw, 2);
        allocator.add(bin, parent_off + 12);
    }
    
    // member: servo5_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint16", js_obj.servo5_raw, 2);
        allocator.add(bin, parent_off + 14);
    }
    
    // member: servo6_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint16", js_obj.servo6_raw, 2);
        allocator.add(bin, parent_off + 16);
    }
    
    // member: servo7_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint16", js_obj.servo7_raw, 2);
        allocator.add(bin, parent_off + 18);
    }
    
    // member: servo8_raw, type: uint16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint16", js_obj.servo8_raw, 2);
        allocator.add(bin, parent_off + 20);
    }
    
}
